src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs

```diff
@@ -119,44 +119,52 @@ private void ValidateWorkloadIdsInput()
-                var packageUrls = GetPackageDownloadUrlsAsync(_workloadIds.Select(id => new WorkloadId(id)), _skipManifestUpdate, _includePreviews).GetAwaiter().GetResult();
+                if (_printDownloadLinkOnly)
+                {
+                    _reporter.WriteLine(string.Format(LocalizableStrings.ResolvingPackageUrls, string.Join(", ", _workloadIds)));
+                    var packageUrls = GetPackageDownloadUrlsAsync(_workloadIds.Select(id => new WorkloadId(id)), _skipManifestUpdate, _includePreviews).GetAwaiter().GetResult();
 
-                _reporter.WriteLine("==allPackageLinksJsonOutputStart==");
-                _reporter.WriteLine(JsonSerializer.Serialize(packageUrls));
-                _reporter.WriteLine("==allPackageLinksJsonOutputEnd==");
-            }
-			else if (!string.IsNullOrWhiteSpace(_downloadToCacheOption))
-            {
-                try
+                    _reporter.WriteLine("==allPackageLinksJsonOutputStart==");
+                    _reporter.WriteLine(JsonSerializer.Serialize(packageUrls));
+                    _reporter.WriteLine("==allPackageLinksJsonOutputEnd==");
+                }
+                else if (!string.IsNullOrWhiteSpace(_downloadToCacheOption))
                 {
-                    DownloadToOfflineCacheAsync(_workloadIds.Select(id => new WorkloadId(id)), new DirectoryPath(_downloadToCacheOption), _skipManifestUpdate, _includePreviews).Wait();
+                    try
+                    {
+                        DownloadToOfflineCacheAsync(_workloadIds.Select(id => new WorkloadId(id)), new DirectoryPath(_downloadToCacheOption), _skipManifestUpdate, _includePreviews).Wait();
+                    }
+                    catch (Exception e)
+                    {
+                        throw new GracefulException(string.Format(LocalizableStrings.WorkloadCacheDownloadFailed, e.Message), e);
+                    }
                 }
-                catch (Exception e)
+                else
                 {
-                    throw new GracefulException(string.Format(LocalizableStrings.WorkloadCacheDownloadFailed, e.Message), e);
+                    try
+                    {
+                        InstallWorkloads(
+                            _workloadIds.Select(id => new WorkloadId(id)),
+                            _skipManifestUpdate,
+                            _includePreviews,
+                            string.IsNullOrWhiteSpace(_fromCacheOption) ? null : new DirectoryPath(_fromCacheOption));
+                    }
+                    catch (Exception e)
+                    {
+                        // Don't show entire stack trace
+                        throw new GracefulException(string.Format(LocalizableStrings.WorkloadInstallationFailed, e.Message), e);
+                    }
                 }
             }
-            else
+            catch (Exception)
```

Comment:
Should this be a catch or a finally? Do we not want to shutdown on success?

Reply:
I cheated a bit here. I didn't want to add shutdown calls in the separate try blocks, but we can

Reply:
> shutdown calls in the separate try blocksshutdown calls in the separate try blocks

let's do that.

A catch all over these doesn't looks right at all. At least there need to be some test